core/examples/http_bench.rs

```diff
@@ -273,17 +304,20 @@ fn op_read(
   let rid = record.arg as u32;
   debug!("read rid={}", rid);
   let mut zero_copy_buf = zero_copy_buf.unwrap();
-  let fut = futures::future::poll_fn(move |cx| {
-    let mut table = lock_resource_table();
-    let stream = table.get_mut::<TcpStream>(rid).ok_or_else(bad_resource)?;
-    let mut f: Box<dyn AsyncRead + Unpin> =
-      Box::new(AsyncRead01CompatExt::compat(&stream.0));
-    AsyncRead::poll_read(Pin::new(&mut f), cx, &mut zero_copy_buf)
-  })
-  .and_then(move |nread| {
+
+  let fut = async move {
+    let nread = {
+      debug!("pre lock read");
+      let mut table = lock_resource_table().await;
+      debug!("post lock read");
+      let stream = table.get_mut::<TcpStream>(rid).ok_or_else(bad_resource)?;
+      let stream = &mut stream.0;
+      stream.read(&mut zero_copy_buf).await?
+    };
```

Comment:
I'm not sure if this bit is correct. I believe that `MutexGuard` is held all the way until `stream.read(&mut zero_copy_buf).await?` completes which is not good. Probably should be implemented the same way as `Accept` above. Rewriting it as standalone future would render `unsafe impl Sync for PinnedBuf {}` not needed